applications/StructuralMechanicsApplication/custom_elements/shell_thick_element_3D3N.cpp

```diff
@@ -0,0 +1,2801 @@
+#define OPT_INTERPOLATE_RESULTS_TO_STANDARD_GAUSS_POINTS(X)
+#else
+#define OPT_INTERPOLATE_RESULTS_TO_STANDARD_GAUSS_POINTS(X) Utilities::InterpToStandardGaussPoints(X)
+#endif // OPT_USES_INTERIOR_GAUSS_POINTS
+#endif // OPT_1_POINT_INTEGRATION
+
+namespace Kratos
+{
+	namespace Utilities
+	{
+		inline void InterpToStandardGaussPoints(double& v1, double& v2,
+			double& v3)
+		{
+			double vg1 = v1;
+			double vg2 = v2;
+			double vg3 = v3;
+#ifdef OPT_AVARAGE_RESULTS
+			v1 = (vg1 + vg2 + vg3) / 3.0;
+			v2 = (vg1 + vg2 + vg3) / 3.0;
+			v3 = (vg1 + vg2 + vg3) / 3.0;
+#else
+			v1 = (2.0*vg1) / 3.0 - vg2 / 3.0 + (2.0*vg3) / 3.0;
+			v2 = (2.0*vg1) / 3.0 + (2.0*vg2) / 3.0 - vg3 / 3.0;
+			v3 = (2.0*vg2) / 3.0 - vg1 / 3.0 + (2.0*vg3) / 3.0;
+#endif // OPT_AVARAGE_RESULTS
+		}
+
+		inline void InterpToStandardGaussPoints(std::vector< double >& v)
+		{
+			if (v.size() != 3) return;
+			InterpToStandardGaussPoints(v[0], v[1], v[2]);
+		}
+
+		inline void InterpToStandardGaussPoints(std::vector< array_1d<double,
+			3> >& v)
+		{
+			if (v.size() != 3) return;
+			for (size_t i = 0; i < 3; i++)
+				InterpToStandardGaussPoints(v[0][i], v[1][i], v[2][i]);
+		}
+
+		inline void InterpToStandardGaussPoints(std::vector< array_1d<double,
+			6> >& v)
+		{
+			if (v.size() != 3) return;
+			for (size_t i = 0; i < 6; i++)
+				InterpToStandardGaussPoints(v[0][i], v[1][i], v[2][i]);
+		}
+
+		inline void InterpToStandardGaussPoints(std::vector< Vector >& v)
+		{
+			if (v.size() != 3) return;
+			size_t ncomp = v[0].size();
+			for (int i = 1; i < 3; i++)
+				if (v[i].size() != ncomp)
+					return;
+			for (size_t i = 0; i < ncomp; i++)
+				InterpToStandardGaussPoints(v[0][i], v[1][i], v[2][i]);
+		}
+
+		inline void InterpToStandardGaussPoints(std::vector< Matrix >& v)
+		{
+			if (v.size() != 3) return;
+			size_t nrows = v[0].size1();
+			size_t ncols = v[0].size2();
+			for (int i = 1; i < 3; i++)
+				if (v[i].size1() != nrows || v[i].size2() != ncols)
+					return;
+			for (size_t i = 0; i < nrows; i++)
+				for (size_t j = 0; j < ncols; j++)
+					InterpToStandardGaussPoints
+					(v[0](i, j), v[1](i, j), v[2](i, j));
+		}
+	}
+
+	// =========================================================================
+	//
+	// CalculationData
+	//
+	// =========================================================================
+
+	ShellThickElement3D3N::CalculationData::CalculationData
+	(const CoordinateTransformationBasePointerType& pCoordinateTransformation,
+		const ProcessInfo& rCurrentProcessInfo)
+		: LCS0(pCoordinateTransformation->CreateReferenceCoordinateSystem())
+		, LCS(pCoordinateTransformation->CreateLocalCoordinateSystem())
+		, CurrentProcessInfo(rCurrentProcessInfo)
+
+	{
+	}
+
+	// =========================================================================
+	//
+	// Class ShellThickElement3D3N
+	//
+	// =========================================================================
+
+	ShellThickElement3D3N::ShellThickElement3D3N(IndexType NewId,
```

Comment:
it would be good if you put at the beginning of the file which is the reference from which you took the implementaiton